### PR TITLE
Context switch support for embedded

### PR DESCRIPTION
**Pull request by xdavidz**
Hi Folks,

Please carefully review this effort :).

Several highlights:
1) I try to summarize the design inside the comment of the code, please let me know if anything is not clean or need more info;
2) I tested on early 104 and 102_dfx, but please feel free to add more test cases, I would love to test as mush as possible.
3) Please let me know if you think anything will be blocked due to this massive change. We aggressively refactored the xclbin related code .

Thanks,
David

### CODE REVIEW

**Comment by abalasa** [comment]
retest this please

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 1]
```diff
@@ -2964,10 +2978,27 @@ sched_init_exec(struct drm_device *drm)
 		exec_core->cq_thread = kthread_run(cq_check, zdev, name);
 	}
 
-	SCHED_DEBUG("<- sched_init_exec\n");
+	SCHED_DEBUG("<- %s\n", __func__);
 	return 0;
 }
 
+static inline void
+fini_configure(struct drm_device *drm)
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/sched_exec.c]: Please comment this function.

File: src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c [hunk 2]
```diff
@@ -0,0 +1,945 @@
...
+	return 0;
+}
+void
+zocl_xclbin_fini(struct drm_zocl_dev *zdev)
+{
+	if (zdev->zdev_xclbin->zx_uuid)
+		vfree(zdev->zdev_xclbin->zx_uuid);
+	if (zdev->zdev_xclbin)
+		vfree(zdev->zdev_xclbin);
+	zdev->zdev_xclbin = NULL;
```
Comment (mamin506) [inline src/runtime_src/core/edge/drm/zocl/zocl_xclbin.c]: In Linux, vfree() will directly return if addr is null. We don't necessary to check if addr is NULL.
This is just a coding style issue. I'm okay with this code.

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 3]
```diff
@@ -25,14 +25,16 @@
...
+#include "zocl_xclbin.h"
 
-/* Including xclbin.h in the scheduler is not good.
- * But let us do this for now. Should add zocl_xclbin.c later
- * and move all XCLBIN related code there.
+/*
+ * The get_cu_support_intr needs this but we will move it to zocl_xclbin.c
+ * in next code change.
  */
 #include "xclbin.h"
```
[the remark this answers is not in the document]
Comment (xdavidz) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: thanks, what I still think is that we should try to make our code:
1) a clean structure that anyone can understand and follow, for example zocl_xclbin.c only does xclbin related things, sched_exec.c only does scheduler related things. As of today, it is ok, but no more exceptions;
2) a clean code style that is ready for linux upstream. the more we make the code ready, the less work we have to do for upstream, so make checkpatch happy is needed. we should "enforce" this rule anyway. :)

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c [hunk 4]
```diff
@@ -25,14 +25,16 @@
...
 
 /* #define SCHED_VERBOSE */
+#define SCHED_VERBOSE
```
[the remark this answers is not in the document]
Comment (xdavidz) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: sorry, removed it.

File: src/runtime_src/core/edge/drm/zocl/sched_exec.c (hunk 1 above)
Comment (xdavidz) [reply src/runtime_src/core/edge/drm/zocl/sched_exec.c]: Added. Please review.

**Pull request merged**